### ISSUE TITLE
[BUG] Poor performance in eclipse (java indexing is very slow)

### ISSUE THREAD

**Issue by rPraml** [projectlombok/lombok#3174]
**Describe the bug**
We noticed that eclipse becomes very slow, especially if you switch git branches and eclipse needs to reindex files

**To Reproduce**
Start eclipse with a huge code base and delete index in Preferences->Java->Rebuild index.
Open the type menu and check how long index rebuild takes (in my case it takes 8 min 45 sec for 5000 files)

I tracked down the problem to `LombokConfiguration.read(ConfigurationKey<T> key, AST<?, ?, ?> ast)` which is called per AST very often. There are about 80 ConfigurationKeys and it seems that the `read` method is called for many keys per AST, respectively java file. 
(I think this depends which lombok features you use)
The BubblingConfigurationResolver adds the second step of complexity. It searches for a lombok.config up to the root directory.

so, if `com.acme.ClassA` is compiled, I see that, `node.getAst().readConfiguration(key)` is called for a lot of configuration keys.
Each call does the following:
- getAbsoluteFileLocation for `com.acme.ClassA` https://github.com/projectlombok/lombok/blob/master/src/core/lombok/core/LombokConfiguration.java#L69
- create a resolver for that location. By default this is a BubblingConfigurationResolver, during construction, a cache lookup is done by file location. It would be better to cache per package here https://github.com/projectlombok/lombok/blob/master/src/core/lombok/core/configuration/FileSystemSourceCache.java#L75
- the BubblingConfigurationResolver then searches for config in the paths
  -  /home/roland/maven/myproject/src/main/java/com/acme/lombok.config
  -  /home/roland/maven/myproject/src/main/java/com/lombok.config
  -  /home/roland/maven/myproject/src/main/java/lombok.config
  - ...
  -  /home/lombok.config
  -  /lombok.config

and this happens for EVERY key that is resolved. So you have per java source files 100s of file io

**Expected behavior**
Be faster. If I add `-Dlombok.disableConfig`, the index rebuild takes only 5 minutes 40 sec

**Version info (please complete the following information):**
 - Lombok version 1.18.22
 - eclipse version: 2022-03 (4.23.0)

**Additional context**
IMHO the LombokConfiguration.read should be refactored to use more caching, as it seems that config key retrievement is one of the hotspots. (I can live with the `-Dlombok.disableConfig` for a while, although it's not ideal)
Cache should hold complete config per package, so that each lombok.config file should be accessed only once.

**Comment by Rawi01** [comment]
Lombok already caches the absolute file location and the config file content (or `null` if it is not present). Yes, it bubbles but it does not touch the filesystem again, at least for 2 seconds. After that time it only reads the file again if the last modified timestamp changes. This usually works because a normal index operation is really fast. On my system it took 7 seconds to index all jars and an unnoticable amount for 300 java files. If your IO is really slow (for example if you use a network storage) it will always check the modified time which might get really expensive.

Possible further optimizations:
- Detect the end of the current index/compilation operation and keep the cached information for the whole operartion.
- Resolve all keys at the same time and store the whole result for a given AST.
- Store the bubble result

You can try to increase the [`RECHECK_FILESYSTEM`](https://github.com/projectlombok/lombok/blob/773b374340db705c3adc9e5e4c4228659c9d311b/src/core/lombok/core/configuration/FileSystemSourceCache.java#L34) constant and check if it reduces the index time.

**Comment by rPraml** [comment]
Yes, there are some caches, but they could be optimized
In LombokConfiguration, there is a call `cache.forUri(sourceLocation)` for each SourceLocation (in my case 5000 files)
As far as I understand, this creates a new object `ConfigurationFile.forDirectory(directory)` for each file in that directory.
So why not caching per package? This would mean we need a `getAbsolutePackageLocation()` in AST and this would save some object creations.

**Comment by Rawi01** [comment]
This would save some objects but leads to an additional IO operation to resolve the parent directory for a file. We can use a second index for that but I'm not sure if that would solve any performance issues.

**Comment by rzwitserloot** [comment]
I'm open for PRs on this but 'performance when using _wildly_ slow underlying I/O' has extremely low priority, so not something we'd ever work on.

**Comment by rPraml** [comment]
I will close this issue, as we do not have the issue any more (we use IntelliJ now). If there are questions, feel free to contact me
